### DISCUSSION TITLE
UART not working after power on by 3V3_EN

### DISCUSSION THREAD

**Discussion by dylshen** [micropython/micropython#10497] [RP2040 / Pico]
Hi, so I've been using pico with MicroPython to be enabled by the 3V3_EN pin and then receive serial commands over UART1. 
Test program is saved as main.py on the pico and runs on power up, as verified by flashing the onboard LED, however the UART is not working unless I use the 'run' through Thonny. 

So to be clear, when using 'run' in Thonny, the device flashes the LED ten times then waits for UART input. Upon receiving input it turns the LED on and off as expected. All good...

When the device is powered on by pulling 3V3_EN high to 5V, the device flashes LED ten times and is running my test program however does not receive the UART commands. 

When powered on by 3V3_EN, ground is common and device is not connected to Thonny.

The same behaviour is shown when using RUN pin. 

Anybody else come across something like this? 

Pico program:

```py
from machine import Pin, UART
import time

Status = machine.Pin(25, Pin.OUT)
LED = machine.Pin(0, Pin.OUT)

uart_pi = UART(1, baudrate=115200, tx=Pin(8), rx=Pin(9))

for i in range(1, 10):
    Status.value(1)
    time.sleep(0.2)
    Status.value(0)
    time.sleep(0.2)
    Status.value(1)

while True:
    rxData = bytes()
    while uart_pi.any() > 0:
        rxData = uart_pi.read()
        decoded_rxData = rxData.decode('utf-8')
        #print(f"{decoded_rxData}")

        if decoded_rxData == "1":
            LED.value(1)
            uart_pi.write(b'LED on\n\r')
            
        if decoded_rxData == "0":
            LED.value(0)
            uart_pi.write(b'LED off\n\r')
```

**Comment by davefes** [comment]
> When the device is powered on by pulling 3V3_EN high to 5V

I have to ask ... is 3V3_EN actually 5V tolerant?  Is this some special function associated with 3V3_EN?

**Reply by dylshen** [reply]
Here is a schematic of the power management on the Pico: 
[![](https://i.stack.imgur.com/zMzT6.png)
](url)

**Comment by robert-hh** [comment]
3V3_En is not a supply pin. If not by USB, you have to supply the board from the VSYS pin. You can then leave 3V3_EN open. It is internally pulled up by a resistor to VSYS. What I assume is, that if you supply at 3V3_EN, then there is some supply through the ESD protection circuit in the regulator.
@davefes 3V3_EN is a pin of the power regulator, and it is 5V tolerant.

**Reply by dylshen** [reply]
I am powering the Pico by USB in both cases of the test. Going to do a bit more digging into it this afternoon and I'll post back with anything of interest.

**Comment by peterhinch** [comment]
The only reason to connect anything to `3V3_EN` is to disable the 3.3V output by forcing the pin low. This might be done in low power applications.

**Reply by dylshen** [reply]
I want to disable the 2040 to ensure that the board rests in a safe state when not in use. I have it connected to relays which control part of a temperature control cabinet with heaters, fans, etc and wanted a failsafe option of powering down the 2040 by 3V3_EN.

**Reply by robert-hh** [reply]
Since you use Thonny, make sure that the complete script was copied to the board and is started by main.py. Thonny seems to hide the actual state of files on the board.

**Reply by dylshen** [reply]
I think I've got it solved. I have added a line at line 8 to re-initialise the UART and now I'm able to send commands after a power cycle by 3V3_EN.

```py
uart_pi = UART(1, baudrate=115200, tx=Pin(8), rx=Pin(9))
uart_pi.init(baudrate=115200, tx=Pin(8), rx=Pin(9))
```

**Comment by cve2022** [comment]
@dylshen 

I have been able to reproduce it here, with Thonny and running from flash, using the reset/EN pin of the Pico instead of your way.

I am using Putty with a FT232 USB-UART converter.

With Thonny, it does run from the editor.

But...

When running from the internal flash, via REPL, importing the module and running the tst_uart_main() method (see below) I get the following error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "tst_uart.py", line 4, in <module>
NameError: name 'machine' isn't defined

```

After removing the "machine." from the two instantiations:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "tst_uart2.py", line 20, in tst_uart_main
UnicodeError: 

```

Thonny does some weird things; it is funny it works at all when running from the editor.

I modified the code a little while testing. Here it is.

```
from machine import Pin, UART
import time

Status = Pin(25, Pin.OUT)
LED = Pin(0, Pin.OUT)
uart_pi = UART(1, baudrate=115200, tx=Pin(8), rx=Pin(9))

for i in range(1, 5):
    Status.value(1)
    time.sleep(0.1)
    Status.value(0)
    time.sleep(0.1)
    Status.value(1)

def tst_uart_main():
    while True:
        rxData = bytes()
        while uart_pi.any() > 0:
            rxData = uart_pi.read()
            decoded_rxData = rxData.decode('utf-8')

            if decoded_rxData == "1":
                LED.value(1)
                uart_pi.write(b'LED on\n\r')
                
            if decoded_rxData == "0":
                LED.value(0)
                uart_pi.write(b'LED off\n\r')
                uart_pi.flush()
```

I think it is not working when running from flash because of the first error above, but it still has the second error that I did not dig into.

**Reply by dylshen** [reply]
Interesting. Perhaps the 'run' in Thonny re-initialises everything? I'm not sure and just throwing the idea out. 

Reason I say that is because I just added: 
```py
uart_pi = UART(1, baudrate=115200, tx=Pin(8), rx=Pin(9))
uart_pi.init(baudrate=115200, tx=Pin(8), rx=Pin(9))
```

And it works. Is this an effective solution for you as well?

**Reply by cve2022** [reply]
Like this?

```
from machine import Pin, UART
import time

Status = machine.Pin(25, Pin.OUT)
LED = machine.Pin(0, Pin.OUT)

uart_pi = UART(1, baudrate=115200, tx=Pin(8), rx=Pin(9))
uart_pi.init(baudrate=115200, tx=Pin(8), rx=Pin(9))

for i in range(1, 10):
    Status.value(1)
    time.sleep(0.2)
    Status.value(0)
    time.sleep(0.2)
    Status.value(1)

while True:
    rxData = bytes()
    while uart_pi.any() > 0:
        rxData = uart_pi.read()
        decoded_rxData = rxData.decode('utf-8')
        #print(f"{decoded_rxData}")

        if decoded_rxData == "1":
            LED.value(1)
            uart_pi.write(b'LED on\n\r')

        if decoded_rxData == "0":
            LED.value(0)
            uart_pi.write(b'LED off\n\r')

```

Saved as x.py in the Pico.

MicroPython v1.19.1-782-g699477d12 on 2022-12-24; Raspberry Pi Pico with RP2040

Type "help()" for more information.

From Thonny's REPL terminal:

```
>>> import x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "x.py", line 4, in <module>
NameError: name 'machine' isn't defined
>>> 
```

FWIW:
```
>>> sys.version
'3.4.0; MicroPython v1.19.1-782-g699477d12 on 2022-12-24'
>>> sys.implementation
(name='micropython', version=(1, 19, 1), _machine='Raspberry Pi Pico with RP2040', _mpy=4358)
>>> 
```

After removing the "machine." from lines 4 and 5, it does work.

**Reply by dylshen** [reply]
I am saving the test program as main.py on the pico so it runs on power up. 

Yes that's where I added the .init() line and I think I'm good with that now. I've tested this in my project and it is now working as I expected it to. 

Ah right I see, I made a mistake when calling the Pin class. So yes, the "machine." on lines 4 and 5 is unnecessary.
